torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -3097,6 +3097,105 @@ bool any_variable_defined(variable_list& variables) {
+  // Each Householder matrix is stored using Householder vectors (columns of 'input') and its scaling factors (scalars of 'tau')
+  // An explicit Householder matrix in code is like this: H_i = eye(input.shape[-1]) - tau[i] * outer(input[:, i], input[:, i])
+  // See "Representation of Orthogonal or Unitary Matrices": https://www.netlib.org/lapack/lug/node128.html
+
+  // In Python code the orgqr operation can be implemented as following:
+  // result = torch.eye(input.shape[-2])
+  // for j in range(tau.shape[-1]):
+  //     v = A[:, j]
+  //     tauj = tau[j]
+  //     v1, v2 = v, v
+  //     result1, result2 = result, result
+  //     v3 = torch.outer(v1, v2)
+  //     v4 = tau[j] * v3
+  //     v5 = result1 @ v4
+  //     result = result2 - v5
+
+  // Differentiating the Python implementation line-by-line gives (sequence of assignments on the right inside the for loop should be applied in reverse order):
+  // result = torch.eye(input.shape[-2])       |  d_input = torch.zeros_like(input)
+  // for j in range(tau.shape[-1]):            |  for j in reversed(range(tau.shape[-1])):
+  //     v = input[:, j]                       |      d_input[:, j] += d_v
+  //     tauj = tau[j]                         |      d_tau[j] += d_tauj.sum(-1).sum(-1)
+  //     v1, v2 = v, v                         |      d_v = d_v1 + d_v2
+  //     result1, result2 = result, result     |      d_result = d_result1 + d_result2
+  //     v3 = torch.outer(v1, v2)              |      d_v1, d_v2 = d_v3 @ v2, v1 @ d_v3
+  //     v4 = tauj * v3                        |      d_tauj, d_v3 = d_v4 * v3, dv4 * tauj
+  //     v5 = result1 @ v4                     |      d_result1, d_v4 = d_v5 @ v4.T, result1.T @ d_v5
+  //     result = result2 - v5                 |      d_result2, d_v5 = d_result, -d_result
+
+  // in order to support rectangular input we need to expand 'input' and 'd_result' to square matrices
+  auto square_shape = input.sizes().vec();
+  square_shape.back() = input.size(-2);
+
+  auto input_ = at::zeros(square_shape, input.options());
+
+  using at::indexing::Slice;
+  using at::indexing::None;
+  input_.index_put_({"...", Slice(), Slice(None, input.size(-1))}, input); // input_[..., :, :input.shape[-1]] = input
+
+  // make sure that all elements of Householder vectors including 0's and 1's are stored explicitly
+  // LAPACK assumes implicitly that the diagonal is filled with ones and the upper triangle is zero
+  input_.tril_(-1);
+  input_.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(1);
+
+  auto d_input = at::zeros_like(input);
+  auto d_tau = at::zeros_like(tau);
+
+  auto d_result = at::zeros(square_shape, grad.options());
+  d_result.index_put_({"...", Slice(), Slice(None, input.size(-1))}, grad); // d_result[..., :, :input.shape[-1]] = grad
+
+  auto start_j = tau.size(-1) - 1;
+  for (int64_t j = start_j; j >= 0; j--) {
+    auto v = input_.index({"...", Slice(), j});
+    auto v1 = v, v2 = v;
```

Comment:
I believe you can replace indexing with ellipsis with `narrow`, should be faster.

Reply:
In my tests, replacing `.index` here and later in code with `.narrow` doesn't make any difference.

Reply:
Of course, compared to matrix products it is almost nothing, so, I guess, it is primarily a taste thingy. I have just never seen `index` used in ATen, so good to know this is yet another tool available :)